Update Ukkonen for v3

Joseph-Edwards opened the pull request:
This PR updates `ukkonen.hpp` for v3

Joseph-Edwards commented:
This is a draft until python bindings are done.

james-d-mitchell commented on include/libsemigroups/ukkonen.hpp:
```diff
@@ -878,343 +1003,664 @@ namespace libsemigroups {
+    //! \throws LibsemigroupsException if `throw_if_not_unique_letters(w)`
+    //! throws.
+    //!
+    //! \sa \ref Ukkonen.throw_if_not_unique_letters.
```
```suggestion
    //! \sa \ref Ukkonen::throw_if_not_unique_letters.
```
Some python slipped back in here

james-d-mitchell commented on include/libsemigroups/ukkonen.hpp:
```diff
@@ -878,343 +1003,664 @@ namespace libsemigroups {
+      add_word(u, w.cbegin(), w.cend());
+    }
+
+    //! \brief \copybrief Ukkonen.add_word_no_checks
```
More python

james-d-mitchell commented on include/libsemigroups/ukkonen.hpp:
```diff
@@ -878,343 +1003,664 @@ namespace libsemigroups {
+    //! \throws LibsemigroupsException if `u.throw_if_not_unique_letters(w, w +
+    //! std::strlen(w))` throws.
+    //!
+    //! \sa \ref Ukkonen.throw_if_not_unique_letters.
```
more python

james-d-mitchell commented on include/libsemigroups/ukkonen.hpp:
```diff
@@ -16,11 +16,6 @@
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 //
 
-// This file contains an implementation of a generalised suffix tree, adapted
```
Note sure why these lines are deleted, can you please undelete them?

james-d-mitchell reviewed (changes requested):
Looks good apart from the deletion of the lines at the start of the file (which should remain for the attribution if nothing else), and the fact that some python seems to have slipped into the references. I spotted some of these but there might be more, can you please check and fix them all? Thanks!

Joseph-Edwards commented:
> Looks good apart from the deletion of the lines at the start of the file (which should remain for the attribution if nothing else), and the fact that some python seems to have slipped into the references. I spotted some of these but there might be more, can you please check and fix them all? Thanks!

All done now. The 'cpplint' job will (hopefully) pass when #615 is merged.

With regard to the deleted comments at the top of the file, I had moved those comments into the doxygen comments a bit further down so it's visible in the doc, but have now added it back as a normal comment at the top just to be doubly safe.

james-d-mitchell commented:
Thanks @Joseph-Edwards can you please rebase onto v3, so that the changes you made for cpplint are picked up?

james-d-mitchell commented:
Great! Thanks @Joseph-Edwards
